Redo autogenerated metadata

annevk opened the pull request:
Fixes #4210.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/w3c/web-platform-tests/6792)
<!-- Reviewable:end -->

w3c-bots commented:
# Build BROKEN

Started: 2017-08-09 14:04:21
Finished: 2017-08-09 14:24:14



View more information about this build on:

- [WPT Results Dashboard](http://pulls.web-platform-tests.org/build/17169)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/262679047)

jgraham commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[tools/manifest/sourcefile.py, line 204 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6792#-Kr6PV_Dzto7YYVvTzLi:-Kr6PV_ErJxo-4nfvsRW:bvqclop) ([raw file](https://github.com/w3c/web-platform-tests/blob/9830c5a474a361e2ee71cd26d8d5f234d081545d/tools/manifest/sourcefile.py#L204)):*
> ```Python
> 
>     @property
>     def name_is_multi_global(self):
> ```

FWIW I would leave `.worker.`, `.window` and `.any.` tests working for now so you don't have to replace all of them at once.

---

*[tools/manifest/sourcefile.py, line 207 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6792#-Kr6PgbAsAfC9T-1i4F2:-Kr6PgbB2tNTD1rrsrmT:b-sq5gox) ([raw file](https://github.com/w3c/web-platform-tests/blob/557156e603e1141ffa61028f26b0e78b21884f9c/tools/manifest/sourcefile.py#L207)):*
> ```Python
>         """Check if the file name matches the conditions for the file to
>         be a multi-global js test file"""
>         return "global" in self.meta_flags and self.ext == ".js"
> ```

`.global.` seems like an odd flag because the technical meaning contradicts the standard meaning here (it makes it sound like the test is somehow global rather than like it runs in multiple globals.

---

*[tools/manifest/sourcefile.py, line 487 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6792#-Kr6Q8xhZdRm78Fwo2w7:-Kr6Q8xi0ZxtEth7xffV:bbjmssp) ([raw file](https://github.com/w3c/web-platform-tests/blob/557156e603e1141ffa61028f26b0e78b21884f9c/tools/manifest/sourcefile.py#L487)):*
> ```Python
>         elif self.name_is_global:
>             rv = TestharnessTest.item_type, []
>             valueTypeSuffixes = []
> ```

Variables should be formatted with underscores i.e. `value_type_suffix`.

---

*[tools/manifest/sourcefile.py, line 491 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6792#-Kr6QlFp19pzrP42dKW2:-Kr6QlFp19pzrP42dKW3:b-xafvsj) ([raw file](https://github.com/w3c/web-platform-tests/blob/557156e603e1141ffa61028f26b0e78b21884f9c/tools/manifest/sourcefile.py#L491)):*
> ```Python
>                 if key == b"global":
>                     valueTypes = value.split(b",")
>                     for valueType in valueTypes:
> ```

So it seems like this would be clearer by building up a set e.g.

    variants = {}
    value_types = sorted(item.strip() for item in value.split(","), key=lambda x:(x[0] == "!", x))
    expansions = {"workers": {"worker", "sharedworker", "serviceworker"}}
    for item in value_types:
        values = expansions.get(item, {item})
        if item.startswith("!"):
            variants -= values
        else:
            variants |= values

    for variant in variants:
        rv[1].append(TestharnessTest(self, replace_end(self.url, ".global.js", ".global.%s.html" % variant), timeout=self.timeout))

I would also put the first bit into its own method. This does mean that window tests aren't just `.html`, but you can special case that in the substitution if you like.

---

*[tools/manifest/sourcefile.py, line 508 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6792#-Kr6QS1kZ8TQ5kb_bbI8:-Kr6QS1l0MLhr7ggmvVE:b-agexu0) ([raw file](https://github.com/w3c/web-platform-tests/blob/557156e603e1141ffa61028f26b0e78b21884f9c/tools/manifest/sourcefile.py#L508)):*
> ```Python
> 
>             if valueTypeSuffixes == []:
>                 raise ValueError("You need to use `// META: global=...` in your resource")
> ```

This type of check should be part of the lint.

---

*[tools/serve/serve.py, line 142 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6792#-Kr6SYkrp19mxEJgsYUd:-Kr6SYkrp19mxEJgsYUe:b-igh5xi) ([raw file](https://github.com/w3c/web-platform-tests/blob/9830c5a474a361e2ee71cd26d8d5f234d081545d/tools/serve/serve.py#L142)):*
> ```Python
> 
> class WindowHandler(HtmlWrapperHandler):
>     path_replace = [(".window.html", ".window.js")]
> ```

Again, I would leave these ones working for now.

---

*[tools/serve/serve.py, line 142 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6792#-Kr6U8PRyK7SqlqnsqIZ:-Kr6U8PRyK7SqlqnsqI_:bn8hzt1) ([raw file](https://github.com/w3c/web-platform-tests/blob/557156e603e1141ffa61028f26b0e78b21884f9c/tools/serve/serve.py#L142)):*
> ```Python
> 
>     def _meta_replacement(self, key, value):
>         if key == b"global":
> ```

I'm not really sure that this isn't just adding unnecessary complexity. In particular you end up duplicating a lot of logic here, so if you really want this I would prefer some way to share the logic with `sourcefile.py.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/web-platform-tests/6792)*
<!-- Sent from Reviewable.io -->

jgraham commented:
You also need to update the lint to allow this metadata. At the moment it will fail because this isn't a meta type it knows about.

annevk commented:
> FWIW I would leave .worker., .window and .any. tests working for now so you don't have to replace all of them at once.

I think it would be better to do that as part of this. Having said that, it would be trivial to keep `window.js` working so I'll add that back in. Would save from adding some metadata. `worker.js` would clash with WorkerJavaScriptHandler.

> .global. seems like an odd flag because the technical meaning contradicts the standard meaning here (it makes it sound like the test is somehow global rather than like it runs in multiple globals.

Okay, in that case I suggest we stick with .any. Shall I name the META variable any too?

> So it seems like this would be clearer by building up a set

With the updates I pushed, do you still think so? (I don't think we should support standalone sharedworker for now btw. There's no demonstrated need. And !window and !worker also don't make sense.)

> I'm not really sure that this isn't just adding unnecessary complexity. In particular you end up duplicating a lot of logic here, so if you really want this I would prefer some way to share the logic with `sourcefile.py.

How is this duplicating logic? This seems like the only place that controls whether something that gets loaded actually ends up making sense.

jgraham commented:
Reviewed 1 of 2 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/web-platform-tests/6792)*
<!-- Sent from Reviewable.io -->

jgraham commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[tools/manifest/sourcefile.py, line 496 at r3](https://reviewable.io:443/reviews/w3c/web-platform-tests/6792#-KrGoPqNm-o9sVinpXN1:-KrGoPqNm-o9sVinpXN2:bait7dm) ([raw file](https://github.com/w3c/web-platform-tests/blob/7c12b328aa8460be557c606b5512bd60d72d31e6/tools/manifest/sourcefile.py#L496)):*
> ```Python
>             for (key, value) in self.script_metadata:
>                 if key == b"global":
>                     global_values = value.split(b",")
> ```

Forgot to mention that you should `.strip()` each item here to allow whitespace in the values. You'll need to update the lint too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/web-platform-tests/6792)*
<!-- Sent from Reviewable.io -->

annevk commented:
> Forgot to mention that you should .strip() each item here to allow whitespace in the values.

Why would we want to allow whitespace?

jgraham commented:
I pushed some proposed changes to https://github.com/w3c/web-platform-tests/tree/meta_global

annevk commented:
I think the main problem is that it requires all tests to be over HTTPS if you want to include a service worker test. As long as HTTPS requires more work to get done I don't think that's a good tradeoff.

jgraham commented:
It doesn't require that.

w3c-bots commented:
# Build BROKEN

Started: 2017-08-10 13:07:29
Finished: 2017-08-10 13:22:07





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/17203)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/263076271)

Ms2ger commented:
What's the status here?

annevk commented:
I want to pick this up again at some point, but I haven't made time recently. If someone else would like to take over to make this happen, be my guest. Whoever picks this up should probably take @jgraham's branch (or the commits there) as a starting point.

foolip commented:
> I want to pick this up again at some point, but I haven't made time recently. If someone else would like to take over to make this happen, be my guest. Whoever picks this up should probably take @jgraham's branch (or the commits there) as a starting point.

Looks like the commit from is actually included in this PR, so I guess just continuing here would work?

jgraham commented:
Yes

foolip commented:
I resolved the conflicts and made sure that `wpt lint` and `wpt run` don't just crash, although `wpt lint` did complain about things that probably need fixing to land this.

@annevk @jgraham is this in a state where careful review is "all" that's needed to get it landed, or was something tricky discovered above that needs fixing first?

foolip commented:
OK, failing lints:
```
css/geometry/DOMMatrix-css-string.worker.js: Support file not in support directory (SUPPORT-WRONG-DIR)
css/geometry/WebKitCSSMatrix.worker.js: Support file not in support directory (SUPPORT-WRONG-DIR)
css/geometry/interfaces.worker.js: Support file not in support directory (SUPPORT-WRONG-DIR)
```

Seems like a regression in what is considered a support file.

annevk commented:
@foolip OP has a list of things that need to be done before this can be merged.

Ms2ger commented:
@jgraham @gsnedders I think this is ready for another round of reviews

Ms2ger commented on tools/manifest/sourcefile.py:
```diff
@@ -49,6 +49,86 @@ def read_script_metadata(f, regexp):
 
 
+_any_variants = {
+    b"window": {"default": True, "suffix": ".any.html"},
```
Thought: maybe we should instead have `b"default": {"longhand": {b"window", b"dedicatedworker"}}`, then you could use
```js
// META: global=!default,blah
```

jgraham commented:
This looks great, thanks for picking it up. I think this is at least somewhat documented, we should update the documentation before landing. Otherwise my only real feedback is that I prefer the design where default can be used explicitly.

---

Reviewed 5 of 5 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[tools/lint/lint.py, line 653 at r7](https://beta.reviewable.io/reviews/w3c/web-platform-tests/6792#-LAYlESL1Q0-dIYPGGo2:-LAYlESNCwqfdqKdSSgU:bxiem9o) ([raw file](https://github.com/w3c/web-platform-tests/blob/5a68a36f3ab3c428799a170c2ad8813905185eac/tools/lint/lint.py#L653)):*
> ```Python
> 
>                         else:
>                             included_variants = set.union(get_default_any_variants(),
> ```

I might move this out of the loop.

---

*[tools/manifest/sourcefile.py, line 53 at r7](https://beta.reviewable.io/reviews/w3c/web-platform-tests/6792#-Kr6IjJTJa_qiEbFOFB7-r7-53:-LAYoYzlC3nHpGxFmeGW:b-vswaky) ([raw file](https://github.com/w3c/web-platform-tests/blob/5a68a36f3ab3c428799a170c2ad8813905185eac/tools/manifest/sourcefile.py#L53)):*
<details><summary><i>Previously, Ms2ger wrote…</i></summary><blockquote>

Thought: maybe we should instead have `b"default": {"longhand": {b"window", b"dedicatedworker"}}`, then you could use
```js
// META: global=!default,blah
```
</blockquote></details>

I like that suggestion better.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/w3c/web-platform-tests/6792#-:-LAYq4ai4vyFjJEKNGE3:b84k5ek)*
<!-- Sent from Reviewable.io -->

Ms2ger commented on tools/manifest/sourcefile.py:
```diff
@@ -49,6 +49,86 @@ def read_script_metadata(f, regexp):
+    globals = get_default_any_variants()
+    global_values = set(item.strip() for item in value.split(b","))
+
+    for item in global_values:
```
Note to self: check if the iteration order here is significant

Ms2ger commented on tools/manifest/sourcefile.py:
```diff
@@ -49,6 +49,86 @@ def read_script_metadata(f, regexp):
+    assert isinstance(value, binary_type), value
+
+    globals = get_default_any_variants()
+    global_values = set(item.strip() for item in value.split(b","))
```
There doesn't seem to be a reason to create the set. (Could also do the `strip()` call inside the loop).

Ms2ger committed:
Extend the multi-global test (.any.js) infrastructure to support other scopes.

This commit is based on ideas and code by Anne van Kesteren and James Graham.

The infrastructure is extended to allow tests to opt-in to being run in shared
workers and service workers, while the default remains as window and dedicated
worker scopes. (This default may be changed in the future.)
